pkg/controller/v1alpha1/routes/route_controller.go

```diff
@@ -266,25 +309,48 @@ func DiscoverPortLabel(ctx context.Context, mLog logr.Logger, workload *unstruct
 	return intstr.IntOrString{}, nil, fmt.Errorf("can't discovery port from workload %v %v.%v and it's child resource, errorList: %v", workload.GetName(), workload.GetAPIVersion(), workload.GetKind(), gatherErrs)
 }
 
-func discoveryFromObject(w *unstructured.Unstructured) (intstr.IntOrString, map[string]string, error) {
-	obj, found, _ := unstructured.NestedMap(w.Object, "spec", "template")
+func discoveryFromPodSpec(w *unstructured.Unstructured, fieldPath string) (intstr.IntOrString, error) {
+	paved := fieldpath.Pave(w.Object)
+	obj, err := paved.GetValue(fieldPath)
+	if err != nil {
+		return intstr.IntOrString{}, err
+	}
+	data, err := json.Marshal(obj)
+	if err != nil {
+		return intstr.IntOrString{}, fmt.Errorf("discovery podSpec from %s in workload %v err %v", fieldPath, w.GetName(), err)
+	}
+	var spec corev1.PodSpec
+	err = json.Unmarshal(data, &spec)
+	if err != nil {
+		return intstr.IntOrString{}, fmt.Errorf("discovery podSpec from %s in workload %v err %v", fieldPath, w.GetName(), err)
+	}
+	port := getFirstPort(spec.Containers)
+	if port == 0 {
+		return intstr.IntOrString{}, fmt.Errorf("no port found in workload %v", w.GetName())
+	}
+	return intstr.FromInt(int(port)), nil
+}
+
+// discoveryFromPodTemplate not only discovery port, will also use labels in podTemplate
+func discoveryFromPodTemplate(w *unstructured.Unstructured, fields ...string) (intstr.IntOrString, map[string]string, error) {
+	obj, found, _ := unstructured.NestedMap(w.Object, fields...)
 	if !found {
 		return intstr.IntOrString{}, nil, fmt.Errorf("not have spec.template in workload %v", w.GetName())
 	}
 	data, err := json.Marshal(obj)
 	if err != nil {
 		return intstr.IntOrString{}, nil, fmt.Errorf("workload %v convert object err %v", w.GetName(), err)
 	}
-	var template corev1.PodTemplate
-	err = json.Unmarshal(data, &template)
+	var spec corev1.PodTemplateSpec
+	err = json.Unmarshal(data, &spec)
 	if err != nil {
 		return intstr.IntOrString{}, nil, fmt.Errorf("workload %v convert object to PodTemplate err %v", w.GetName(), err)
 	}
-	port := getFirstPort(template.Template.Spec.Containers)
+	port := getFirstPort(spec.Spec.Containers)
 	if port == 0 {
 		return intstr.IntOrString{}, nil, fmt.Errorf("no port found in workload %v", w.GetName())
 	}
-	return intstr.FromInt(int(port)), template.Labels, nil
+	return intstr.FromInt(int(port)), spec.Labels, nil
 }
 
 func getFirstPort(cs []corev1.Container) int32 {
```

Comment:
Just curious why we just get the first port of the container. Is it because KubeVela just supports only one port currently?

Reply:
yes, I don't have way to handle multiple port yet